src/System.Private.ServiceModel/tools/test/SelfHostWcfService/CertificateResources/EndCertificateResource.cs

```diff
@@ -0,0 +1,85 @@
+using System.Text;
+using WcfTestBridgeCommon;
+
+namespace WcfService.CertificateResources
+{
+    internal abstract class EndCertificateResource : CertificateResource
+    {
+        public override ResourceResponse Get(ResourceRequestContext context)
+        {
+            string thumbprint;
+            bool thumbprintPresent = context.Properties.TryGetValue(thumbprintKeyName, out thumbprint) && !string.IsNullOrWhiteSpace(thumbprint);
+
+            string subject;
+            bool subjectPresent = context.Properties.TryGetValue(subjectKeyName, out subject) && !string.IsNullOrWhiteSpace(subject);
+
+            ResourceResponse response = new ResourceResponse();
+
+            // if no subject and no thumbprint parameter provided, provide a list of certs already PUT to this resource 
+            if (!thumbprintPresent && !subjectPresent)
+            {
+                string retVal = string.Empty;
+                string[] subjects;
+                string[] thumbprints;
+
+                lock (s_certificateResourceLock)
+                {
+                    int certNum = s_createdCertsBySubject.Count;
+                    subjects = new string[certNum];
+                    thumbprints = new string[certNum];
+
+                    foreach (var keyVal in s_createdCertsBySubject)
+                    {
+                        --certNum;
+                        subjects[certNum] = keyVal.Key;
+                        thumbprints[certNum] = keyVal.Value.Thumbprint;
+                    }
+                }
+
+                response.Properties.Add(subjectsKeyName, string.Join(",", subjects));
+                response.Properties.Add(thumbprintsKeyName, string.Join(",", thumbprints));
+                return response;
+            }
+            else
+            {
+                // Otherwise, check on the creation state given the certificate thumbprint or subject
+                // thumbprint is given priority if present
+
+                X509Certificate2 certificate = null;
+                bool certHasBeenCreated = false;
+
+                lock (s_certificateResourceLock)
+                {
+                    if (thumbprintPresent)
+                    {
+                        certHasBeenCreated = s_createdCertsByThumbprint.TryGetValue(thumbprint, out certificate);
+                    }
+                    else if (subjectPresent)
+                    {
+                        certHasBeenCreated = s_createdCertsBySubject.TryGetValue(subject, out certificate);
+                    }
+                }
+
+                if (certHasBeenCreated)
+                {
+                    response.Properties.Add(thumbprintKeyName, certificate.Thumbprint);
+                    response.Properties.Add(certificateKeyName, Convert.ToBase64String(certificate.RawData));
+                }
+                else
+                {
+                    response.Properties.Add(thumbprintKeyName, string.Empty);
```

Comment:
A GET for an unknown cert thumbprint/subject shouldn't happen legally. 

IMO, at most, this would be a 404 if we provided a valid thumbprint or subject but it wasn't found. 400 if the thumbprint or subject was invalid data (e.g., blank or had invalid characters or invalid length?)